MSSQL exclude schema name when diff view definition

MatrixDai opened the pull request:
…ancies.

<!--- This environment context section helps us quickly review your PR. 
      Please take a minute to fill-out this information. -->
## Environment

**Liquibase Version**: 4.4

**Liquibase Integration & Version**: CLI

**Liquibase Extension(s) & Version**: NA

**Database Vendor & Version**: MSSQL

**Operating System Type & Version**: All

## Pull Request Type

<!--- What types of changes does your code introduce?
      Put an `x` in all the boxes that apply: 
      If this PR fixes an existing GH issue, edit the next line to add "closes #XXXX" to auto-link.
      If this PR fixes an existing CORE Jira issue, note that as well, although there will be no auto-linking. -->
- [ ] Bug fix (non-breaking change which fixes an issue.)
- [x] Enhancement/New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
Liquibase report schema diff discrepancy due to "schema" (e.g: `dbo`) are included in the view definition.


## Steps To Reproduce
Create the following view in 2 MSSQL database (all in default dbo schema):


```
SET ANSI_NULLS ON
GO
SET QUOTED_IDENTIFIER ON
GO
create view [dbo].[view_demo] with schemabinding as select col1 as col1, col2 as col2 from dbo.table_demo where age > 1
GO
```

Execute liquibase diff command result in the following discrepancy:


```
Changed View(s): 
     view_demo
          definition changed from 'create view view_demo with schemabinding as select col1 as col1, col2 as col2 from dbo.table_demo where age > 1' to 'create view [dbo].[view_demo] with schemabinding as select col1 as col1, col2 as col2 from dbo.table_demo where age > 1'
```

## Actual Behavior
Liquibase report view changed.

## Expected/Desired Behavior
Liquibase should not report such discrepancy.

## Screenshots (if appropriate)
NA

## Additional Context
The 2 views are created on 2 different server version, maybe this is the root cause?

Server1: `13.00.4422`
Server2: `13.00.5026`

## Fast Track PR Acceptance Checklist:
<!--- Completing these speeds up the acceptance of your pull request -->
<!--- Put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, just ask us in a comment. We're here to help! -->
- [x] Build is successful and all new and existing tests pass
- [ ] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
- [ ] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
- [ ] Added [Test Harness Test(s)](https://github.com/liquibase/liquibase-test-harness/pulls)
- [ ] Documentation Updated

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

molivasdat commented:
Hi @MatrixDai Thanks for creating this enhancement request.

A member of the Liquibase team will take a look at your contribution and may suggest: 

- Changes based on [guidelines for PR submission](https://liquibase.jira.com/wiki/spaces/LB/pages/1274904896/How+to+Contribute+Code+to+Liquibase+Core)
- Additional tests ([unit ](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)and/or [integration](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)) and/or [Test Harness Tests](https://github.com/liquibase/liquibase-test-harness/pulls)
- Provide other feedback, like requesting changes. 

The PR will be prioritized according to our internal development and testing capacity.

We’ll let you know when it’s ready to move to the next step or if any changes are needed.

MatrixDai committed:
Avoid schema name in 'create view' definition causing MSSQL view diff discrepancies.

- Add MSSQLDatabase import to ViewSnapshotGenerator
- Strip optional schema prefix from CREATE VIEW definition for MSSQL (ported to liquibase-standard module)

MatrixDai committed:
Improve MSSQL view schema-stripping regex robustness.

- Use Pattern.quote() to safely handle schema names with regex metacharacters
- Broaden view name match from [a-z][a-z0-9_$#@]* to [^\]\s]+ to cover identifiers starting with _, #, @ or containing hyphens and unicode
- Add null guards for view.getSchema() and view.getSchema().getName()

filipelautert reviewed (commented):
Hey, nice fix for the schema-prefix diff noise!

One thing to consider: the regex `\[?([^\]\s]+)\]?` also strips the `[]` brackets from the view name, not just the schema prefix. So `create view [dbo].[view_demo]` becomes `create view view_demo` instead of `create view [view_demo]`.

For the reported issue this works fine (both sides end up unbracketed), but if one server returns `create view [view_demo]` (brackets, no schema) while the other returns `create view [dbo].[view_demo]`, you'd still get a false diff (`[view_demo]` vs `view_demo`).

Worth considering whether the replacement should preserve the view name's original brackets, something like:

```java
definition = definition.replaceFirst("(?i)(create\\s+view\\s+)\\[?"
        + Pattern.quote(view.getSchema().getName())
        + "\\]?\\.(?=\\[?[^\\]\\s]+\\]?)", "$1");
```

This strips only `[schema].` and leaves the view name (with or without brackets) untouched.

MatrixDai committed:
Fix MSSQL view schema-stripping regex to preserve view name brackets.

- Use lookahead instead of capture group for view name so brackets are not stripped
- Prevents false diff when one definition has brackets and another does not

rberezen reviewed (commented):
@MatrixDai 

Tested on MSSQL. The PR correctly strips the schema prefix from `CREATE VIEW` definitions, but a false diff remains due to **bracket inconsistency on the view name**.

After stripping, one DB has `CREATE VIEW view_demo ...` and the other has `CREATE VIEW [view_demo] ...`. The lookahead preserves whatever brackets are present.

The regex should also normalize brackets on the view name. Change from:
```java
definition = definition.replaceFirst("(?i)(create\s+view\s+)\[?"
    + Pattern.quote(view.getSchema().getName())
    + "\]?\.(?=\[?[^\]\s]+\]?)", "$1");
```
To:
```java
definition = definition.replaceFirst("(?i)(create\s+view\s+)\[?"
    + Pattern.quote(view.getSchema().getName())
    + "\]?\.\[?([^\]\s]+)\]?", "$1$2");
```
This captures the view name without brackets and outputs it plain, eliminating the remaining false diff.

MatrixDai committed:
Fix MSSQL view false diff: two-pass schema/bracket normalization

- Pass 1 strips [schema]. prefix when present (existing fix)
- Pass 2 strips brackets from the view name itself
- Covers the case where OBJECT_DEFINITION returns [view_name] without
  schema prefix on one server and [schema].[view_name] on another

MatrixDai commented:
`OBJECT_DEFINITION()` may return `CREATE VIEW` headers in different formats depending on the server.
Two passes are applied to normalize them all to the same form:

| Input from OBJECT_DEFINITION | After pass 1 (strip schema prefix) | After pass 2 (strip view name brackets) |
|---|---|---|
| `CREATE VIEW [dbo].[view_demo]` | `CREATE VIEW [view_demo]` | `CREATE VIEW view_demo` |
| `CREATE VIEW [dbo].view_demo` | `CREATE VIEW view_demo` | `CREATE VIEW view_demo` (no-op) |
| `CREATE VIEW [view_demo]` (no schema) | no-op | `CREATE VIEW view_demo` |
| `CREATE VIEW view_demo` (no schema, no brackets) | no-op | no-op → `CREATE VIEW view_demo` |

cc @rberezen @filipelautert

rberezen reviewed (commented):
Tested the updated two-pass fix on MSSQL with the same scenario.

**Before:** `Changed View(s): view_demo` (brackets still differed)
**After:** `Changed View(s): NONE` ✓

The two-pass approach correctly normalizes both schema prefix and view name brackets. Thanks for the quick turnaround, @MatrixDai!
